src/tip-opengov.ts

```diff
@@ -29,40 +31,73 @@ export async function tipOpenGov(opts: {
   if ("error" in track) {
     return { success: false, errorMessage: track.error };
   }
+  const contributorAddress = contributor.account.address;
 
   const proposalTx = api.tx.utility.batch([
     api.tx.system.remark(formatReason(tipRequest)),
-    api.tx.treasury.spend(track.value.toString(), contributor.account.address),
+    api.tx.treasury.spend(track.value.toString(), contributorAddress),
   ]);
   const encodedProposal = proposalTx.method.toHex();
   const proposalHash = blake2AsHex(encodedProposal);
+  const encodedLength = Math.ceil((encodedProposal.length - 2) / 2);
 
-  const preimage_unsub = await api.tx.preimage
-    .notePreimage(encodedProposal)
-    .signAndSend(botTipAccount, { nonce: -1 }, (result) => {
-      if (result.status.isInBlock) {
-        bot.log(`Preimage Upload included at blockHash ${result.status.asInBlock.toString()}`);
-      } else if (result.status.isFinalized) {
-        bot.log(`Preimage Upload finalized at blockHash ${result.status.asFinalized.toString()}`);
-        preimage_unsub();
-      }
-    });
+  return await new Promise(async (resolve, reject) => {
+    // create a preimage from opengov with the encodedProposal above
+    const preimageUnsubscribe = await api.tx.preimage
+      .notePreimage(encodedProposal)
+      .signAndSend(botTipAccount, { nonce: -1 }, async (result) => {
+        await signAndSendCallback(bot, contributor.account, "preimage", preimageUnsubscribe, result)
+          .then(async () => {
+            const readPreimage = await api.query.preimage.statusFor(proposalHash);
 
-  const referenda_unsub = await api.tx.referenda
-    .submit(
-      // TODO: There should be a way to set those types properly.
-      { Origins: track.track } as any, // eslint-disable-line
-      { Lookup: { hash: proposalHash, length: proposalTx.length - 1 } },
-      { after: 10 } as any, // eslint-disable-line
-    )
-    .signAndSend(botTipAccount, { nonce: -1 }, (result) => {
-      if (result.status.isInBlock) {
-        bot.log(`Tip referendum included at blockHash ${result.status.asInBlock.toString()}`);
-      } else if (result.status.isFinalized) {
-        bot.log(`Tip referendum finalized at blockHash ${result.status.asFinalized.toString()}`);
-        referenda_unsub();
-      }
-    });
+            if (readPreimage.isEmpty) {
```

Comment:
Asking for understanding:
At this point, we made sure that the image submission transaction is `isInBlock` AND `isFinalized`.
How can it be that it's empty then?

Reply:
it can be, if account had insufficient amount of money, then the preimage will still be inBlock or Finalized, but not by fact be on-chain.

Reply:
![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/02d606cb-e6a1-44c6-8d8d-9a5c17b694fd)
here i have 0,4 UNIT transferable, but for creating a preimage i will need 1,3

this is what happens with creating a preimage 
![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/c39527bb-8330-41cc-bee9-8f4b2cd343a5)

![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/28e5b37b-1725-4205-b431-35c4fcf2effe)
In logs though all good, finalized, we proceeding to craetee a proposal. But there's no preimage. So it ends up with a proposal, without an "action" to execute after it's Voted and enaction time comes

Here's what happens with validation
![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/1ee8f773-04f9-4ffd-b48a-c1522051804c)
![image](https://github.com/paritytech/substrate-tip-bot/assets/1177472/2491c15d-e296-493e-a459-99babc1933d9)